0013-RS-pubsub-subscription-streaming.md

```diff
@@ -0,0 +1,159 @@
+# PubSub Subscription Streaming
+
+* Author(s): @joshvanl
+* State: Ready for Implementation
+* Updated: 2024-03-05
+
+## Overview
+
+This is a design proposal to implement a new Dapr runtime gRPC and HTTP API for subscription streaming.
+This new gRPC & HTTP API will allow an application to subscribe to a PubSub topic and receive messages through this RPC.
+Applications will be able to dynamically subscribe and unsubscribe to topics, and receive messages without opening a port to receive incoming traffic from Dapr.
+
+## Background
+
+Dapr supports applications subscribing to PubSub topic events.
+These subscriptions can be configured either:
+- `programmatically` by returning the subscription config on the app channel server on app health ready, or
+- `declaratively` via Subscription yaml manifests in Self-Hosted or Kubernetes mode.
+
+Today, it is not possible to dynamically update the subscription list without restarting Daprd, though hot reloading for Subscription manifests is [planned](https://github.com/dapr/dapr/issues/7139).
+It is common for users to want to dynamically subscribe and unsubscribe to topics inside their applications based on runtime conditions.
+In the cases where Dapr is not running as a sidecar, users often do not want to open a public port or create a tunnel in order to receive PubSub messages from Dapr.
+
+A streaming Subscription API will allow applications to dynamically subscribe to PubSub topics and receive messages without opening a port to receive incoming traffic from Dapr.
+
+## Expectations and alternatives
+
+This proposal outlines the gRPC & HTTP streaming API for subscribing to PubSub topics.
+This proposal does _not_ address any hot-reloading functionality to the existing programmatic or declarative subscription configuration.
+Using a gRPC streaming API is the most natural fit for this feature, as it allows for first class long-lived bi-directional connections to Dapr to receive messages.
+A supplementary WebSocket based HTTP API is useful for applications which do not have a gRPC client available or HTTP WebSockets are preferred.
+These messages are typed RPC giving the best UX in each SDK.
+Once implemented, this feature will need to be implemented in all Dapr SDKs.
+
+## Solution
+
+### gRPC
+
+Rough gRPC PoC implementation: https://github.com/dapr/dapr/commit/ed40c95d11b78ab9a36a4a8f755cf89336ae5a05
+
+The Dapr runtime gRPC server will implement the following new RPC and messages:
+
+```proto
+service Dapr {
+  // SubscribeTopicEvents subscribes to a PubSub topic and receives topic events
+  // from it.
+  rpc SubscribeTopicEvents(stream SubscribeTopicEventsRequest) returns (stream TopicEventRequest) {}
```

Comment:
```suggestion
  rpc SubscribeTopicEvents(stream SubscribeTopicEventsRequestAck) returns (stream TopicEventRequest) {}
```

Reply:
`Ack` is not appropriate at the first message on this stream will be the initial request, not an acknowledgment.

Reply:
I was more looking for a name which covers both messages.